librz/analysis/arch/8051/8051_il.c

```diff
@@ -0,0 +1,455 @@
+	[I8051_Z] = "z",
+	[I8051_OV] = "ov",
+	[I8051_RS0] = "rs0",
+	[I8051_RS1] = "rs1",
+	[I8051_N] = "n",
+	[I8051_AC] = "ac",
+	[I8051_CY] = "cy",
+	[I8051_A] = "a",
+	[I8051_B] = "b",
+	NULL,
+};
+
+static const bool i8051_register_is_bit[0xff] = {
+	false,
+	[I8051_Z] = true,
+	[I8051_OV] = true,
+	[I8051_RS0] = true,
+	[I8051_RS1] = true,
+	[I8051_N] = true,
+	[I8051_AC] = true,
+	[I8051_CY] = true,
+};
+
+#define ADD3(a, b, c) ADD(a, ADD(b, c))
+#define SUB3(a, b, c) SUB(SUB(a, b), c)
+#define AND3(a, b, c) AND(a, AND(b, c))
+
+#define LO_BYTE(x) CAST(8, IL_FALSE, x)
+#define HI_BYTE(x) CAST(8, IL_FALSE, SHIFTR0(x, U16(8)))
+
+#define VARG_8051(x) VARG(i8051_registers_str[x])
+#define VAL_A        VARG_8051(I8051_A)
+#define VAL_B        VARG_8051(I8051_B)
+#define VAL_CY       VARG_8051(I8051_CY)
+#define VAL_CY8      BOOL_TO_BV(VARG_8051(I8051_CY), 8)
+
+static RzILOpPure *val_register(I8051OpAddressing *a) {
+	switch (a->d.reg) {
+	case I8051_PC:
+		return U16(a->pc);
+		//	case I8051_CY:
+		//	case I8051_AC:
+		//	case I8051_OV:
+		//		return BOOL_TO_BV(VARG(i8051_registers_str[a->d.reg]), 8);
+	default:
+		return VARG(i8051_registers_str[a->d.reg]);
+	}
+}
+
+static inline RzILOpPure *bv_i(RzILOpBitVector *v, ut8 i) {
+	return NON_ZERO(LOGAND(v, U8(1 << i)));
+}
+
+static inline RzILOpPure *bv_i_bv(RzILOpBitVector *v, ut8 i) {
+	return SHIFTR0(LOGAND(v, U8(1 << i)), U8(i));
+}
+
+static inline RzILOpEffect *set_reg(I8051Registers reg, RzILOpBitVector *v) {
+	return SETG(i8051_registers_str[reg], v);
+}
+
+static inline RzILOpEffect *set_a(RzILOpBitVector *v) {
+	return set_reg(I8051_A, v);
+}
+
+static inline RzILOpEffect *set_b(RzILOpBitVector *v) {
+	return set_reg(I8051_B, v);
+}
+
+static inline RzILOpEffect *set_cy(RzILOpBitVector *v) {
+	return set_reg(I8051_CY, v);
+}
+
+static RzILOpPure *get_any(I8051OpAddressing *a) {
+	switch (a->mode) {
+	case I8051_ADDRESSING_REGISTER:
+		return val_register(a);
+	case I8051_ADDRESSING_DIRECT:
+		return LOAD(U16(a->d.addr));
+	case I8051_ADDRESSING_INDIRECT:
+		return LOAD(CAST(16, IL_FALSE, get_any(a->d.indirect)));
+	case I8051_ADDRESSING_IMMEDIATE:
+		return U8(a->d.constant & 0xff);
+	case I8051_ADDRESSING_IMMEDIATE16:
+		return U16(a->d.constant);
+	case I8051_ADDRESSING_RELATIVE:
+		return ADD(U16(a->pc), U16(a->d.constant));
+	case I8051_ADDRESSING_ABSOLUTE:
+	case I8051_ADDRESSING_LONG:
+		return U16(a->d.addr16);
+	case I8051_ADDRESSING_INDEXED:
+		return ADD(CAST(16, IL_FALSE, LOAD(CAST(16, IL_FALSE, VARG_8051(I8051_A)))),
+			val_register(a));
+	case I8051_ADDRESSING_BIT:
+		return NON_ZERO(LOADW(1, U16(a->d.addr)));
+	default:
+		rz_warn_if_reached();
+		return NULL;
+	}
+}
+static RzILOpEffect *set_any(I8051OpAddressing *a, RzILOpPure *v) {
+	switch (a->mode) {
+	case I8051_ADDRESSING_REGISTER:
+		return set_reg(a->d.reg, v);
+	case I8051_ADDRESSING_BIT: {
+		return STOREW(U16(a->d.addr), BOOL_TO_BV(v, 1));
+	}
+	case I8051_ADDRESSING_DIRECT:
+		return STORE(U16(a->d.addr), v);
```

Comment:
Instead of duplicating the address calculation logic, just use `return STOREW(get_any(a), v)`.

Reply:
But it should not be equivalent